src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+(ns metaprob.autodiff
+  (:refer-clojure :exclude [+ * - / == < > <= >= zero? number? pos? neg?]))
+
+;; Dual numbers represent computations dependent on a value with respect to which we are differentiating.
+;; derivative contains the current derivative; value contains the current value; and tag is a special tag
+;; that helps distinguish between multiple "levels" of nested differentiations.
+(defn make-dual-number [tag value derivative]
+  {:tag tag :value value :derivative derivative})
+
+(declare * + - /)
+
+(def non-dual? clojure.core/number?)
```

Comment:
New to Clojure here :)  Personally I would lean more defensive here, something like
```
(defn dual-number? [x]
  (every? (fn [k] (contains? x k))
          [:value :derivative :tag]))
```
Having a public function with this name sounds ([spiritual](https://en.wikipedia.org/wiki/Nondualism) :slightly_smiling_face: and) a bit prone to being accidentally used incorrectly, since the caller may not realize that `non-dual?` is different from "not `dual?`" (they differ in how they treat things that are neither numbers nor dual numbers).  Wdyt?

Reply:
I agree the name isn't great (and also, there's no reason it needs to be public -- we could make it private). We could just use `clojure.core/number?` -- it just seemed a bit long. Semantically, the places that use this actually do want to do exactly what `clojure.core/number?` does.

Reply:
Maybe we could make it a private function called `bare-number?` or `number?` within this namespace?